src/libclipper/test/json_util_test.cpp

```diff
@@ -0,0 +1,207 @@
+      "{\"double_val\":0.3,\"double_array\":[1.4,2.23,3.243242,0.3223424]}";
+
+  rapidjson::Document d;
+  d.SetObject();
+
+  double old_double_val = 0.5;
+  double new_double_val = 0.3;
+  std::vector<double> old_double_array = {1.5,2.23,3.243242,0.3223424};
+  std::vector<double> new_double_array = {1.4,2.23,3.243242,0.3223424};
+  add_double(d, "double_val", old_double_val);
+  add_double(d, "double_val", new_double_val);
+  add_double_array(d, "double_array", old_double_array);
+  add_double_array(d, "double_array", new_double_array);
+  std::string output_json = to_json_string(d);
+  EXPECT_EQ(output_json, expected_json);
+}
+
+TEST(JsonUtilTests, TestAddEmptyValues) {
+  std::string expected_json =
+      "{\"empty_string\":\"\",\"empty_int_array\":[]}";
+
+  rapidjson::Document d;
+  d.SetObject();
+
+  std::vector<int> empty_int_array = {};
+  add_string(d, "empty_string", "");
+  add_int_array(d, "empty_int_array", empty_int_array);
+}
+
+/* Test JSON deserialization utilities */
+TEST(JsonUtilTests, TestDoubleIntFormatting) {
+  // Numbers require a decimal point to be a float, and cannot have a
+  //  decimal point as an integer.
+  std::string source_json =
+      "{\"number_with_decimal\": 0.3, \"number_without_decimal\": 5}";
+  rapidjson::Document d;
+  parse_json(source_json, d);
+  EXPECT_EQ(get_double(d, "number_with_decimal"), 0.3);
+  ASSERT_THROW(get_int(d, "number_with_decimal"), json_semantic_error);
+  ASSERT_THROW(get_double(d, "number_without_decimal"), json_semantic_error);
+  EXPECT_EQ(get_int(d, "number_without_decimal"), 5);
+}
+
+TEST(JsonUtilTests, TestArrayTypeMismatch) {
+  // JSON arrays may contain items of different types so the array
+  //  parsing methods should throw an exception if this is the case.
+  std::string source_json =
+      "{\"double_array_with_int\": [1.2, 3.4, 5, 6.7],"
+       "\"int_array_with_string\": [1, 2, 3, \"4\"],"
+       "\"string_array_with_bool\": [\"test1\", \"phrase2\", true]}";
+
+  rapidjson::Document d;
+  parse_json(source_json, d);
+  ASSERT_THROW(
+        get_double_array(d, "double_array_with_int"),
+        json_semantic_error);
+  ASSERT_THROW(
+        get_int_array(d, "int_array_with_string"),
+        json_semantic_error);
+  ASSERT_THROW(
+        get_string_array(d, "string_array_with_bool"),
+        json_semantic_error);
+}
+
+TEST(JsonUtilTests, TestEmptyDocumentToString) {
+  rapidjson::Document d;
+  EXPECT_EQ("null", to_json_string(d));
+  d.SetObject();
+  EXPECT_EQ("{}", to_json_string(d));
+  d.SetArray();
+  EXPECT_EQ("[]", to_json_string(d));
+  d.SetDouble(0.3);
+  EXPECT_EQ("0.3", to_json_string(d));
+  d.SetBool(true);
+  EXPECT_EQ("true", to_json_string(d));
+}
+
+TEST(JsonUtilTests, TestAddToNonObject) {
+  rapidjson::Document d;
+  d.SetArray();
+  ASSERT_THROW(add_int(d, "int_val", 3), json_semantic_error);
+}
+
+TEST(JsonUtilTests, TestParseCandidateModels) {
+  // Semantic checking to assert that model_version > 0?
```

Comment:
Decided to remove the comment. For now, we'll only handle formatting errors and basic semantic errors (e.g. type mismatches).